include/envoy/matcher/matcher.h

```diff
@@ -0,0 +1,200 @@
+// from the beginning every time. At some point we might support resuming for any node in the match
+// tree: this requires some careful handling of tracking which on_no_match to use should we fail to
+// match.
+//
+// All the matching is performed on strings: a DataInput<DataType> is used to extract a specific
+// string from an instance of DataType, while an InputMatcher is used to determine whether the
+// extracted string is a match.
+//
+// For example, DataType might be the type HttpDataInput, allowing
+// for the use of HttpRequestHeaders : DataInput<HttpDataInput>, which is configured with the name
+// of the header to extract from the request headers.
+//
+// In cases where the data to match on becomes available over time, this would be fed into the
+// DataType over time, allowing matching to be re-attempted as more data is made available. As such
+// whenever we extract data from a DataInput, we make note of whether the data might change and
+// pause matching until we either match or have all the data. It would then fall on the caller to
+// both provide more information to the DataType and to resume matching.
+template <class DataType> class MatchTree;
+
+template <class DataType> using MatchTreeSharedPtr = std::shared_ptr<MatchTree<DataType>>;
+
+/**
+ * Action provides the interface for actions to perform when a match occurs. It provides no
+ * functions, as implementors are expected to downcast this to a more specific action.
+ */
+class Action {
+public:
+  virtual ~Action() = default;
+
+  /**
+   * The underlying type of this action. This can be used to determine which
+   * type this action is before attempting to cast it.
+   */
+  virtual absl::string_view typeUrl() const PURE;
+
+  /**
+   * Helper to convert this action to its underlying type.
+   */
+  template <class T> T& getTyped() {
+    ASSERT(dynamic_cast<T*>(this) != nullptr);
+    return static_cast<T&>(*this);
+  }
+};
+
+using ActionPtr = std::unique_ptr<Action>;
+using ActionFactoryCb = std::function<ActionPtr()>;
+
+class ActionFactory : public Config::TypedFactory {
+public:
+  virtual ActionFactoryCb createActionFactoryCb(const Protobuf::Message& config) PURE;
+
+  std::string category() const override { return "envoy.matching.action"; }
+};
+
+// On match, we either return the action to perform or another match tree to match against.
+template <class DataType> struct OnMatch {
+  const ActionFactoryCb action_cb_;
+  const MatchTreeSharedPtr<DataType> matcher_;
+};
+
+/**
+ * MatchTree provides the interface for performing matches against the data provided by DataType.
+ */
+template <class DataType> class MatchTree {
+public:
+  virtual ~MatchTree() = default;
+
+  // This encodes three states:
+  // Not enough data to complete the match: {false, {}}
+  // Completed the match, no match: {true, {}}
+  // Completed the match, match: {true, on_match}
+  struct MatchResult {
+    const bool match_completed_;
```

Comment:
nit: I think this code and callers would be much easier to read if you use a two state enum here. Possibly other places you are returning a bool, but up to you.